lib/hierarchical.h

```diff
@@ -465,6 +477,117 @@ class HierarchicalRadiosity {
         }
     }
 
+    /**
+     * Refine all links in all nodes.
+     * @return true if links have been refined.
+     */
+    bool refine_links() {
+        bool done = true;
+        for (auto& p : nodes_) {
```

Comment:
Hier würde ich `p` in `root` umbenennen. So ist das auch an den anderen Stellen.